Common/Console.ps1

```diff
@@ -0,0 +1,101 @@
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+#----------------------------------------------------------------------------------------------------------------------
+Set-StrictMode -Version Latest
+$ErrorActionPreference = 'Stop'
+
+$CSharpCode = @'
+private const uint GENERIC_READ = 0x80000000;
+private const uint GENERIC_WRITE = 0x40000000;
+private const uint FILE_SHARE_READ = 0x00000001;
+private const uint FILE_SHARE_WRITE = 0x00000002;
+private const uint OPEN_EXISTING = 3;
+
+[DllImport("kernel32.dll", SetLastError = true, CharSet=CharSet.Unicode, ExactSpelling = true)]
+private static extern SafeFileHandle CreateFileW(string fileName, uint desiredAccess, uint shareMode, IntPtr securityAttributes, uint creationDisposition, uint flagsAndAttributes, IntPtr templateFile);
+
+[DllImport("kernel32.dll", SetLastError = true, ExactSpelling = true)]
+public static extern uint GetConsoleMode(SafeFileHandle consoleHandle, out uint mode);
+
+public static SafeFileHandle GetConIn()
+{
+    SafeFileHandle fileHandle = CreateFileW("CONIN$", GENERIC_READ, FILE_SHARE_READ, IntPtr.Zero, OPEN_EXISTING, 0, IntPtr.Zero);
+    if (fileHandle.IsInvalid)
+    {
+        throw new System.ComponentModel.Win32Exception();
+    }
+    return fileHandle;
+}
+
+public static SafeFileHandle GetConOut()
+{
+    SafeFileHandle fileHandle = CreateFileW("CONOUT$", GENERIC_READ | GENERIC_WRITE, FILE_SHARE_WRITE, IntPtr.Zero, OPEN_EXISTING, 0, IntPtr.Zero);
+    if (fileHandle.IsInvalid)
+    {
+        throw new System.ComponentModel.Win32Exception();
+    }
+    return fileHandle;
+}
+'@
+
+$NativeMethods = Add-Type -Language CSharp -MemberDefinition $CSharpCode -Name NativeMethods -Namespace Console -PassThru -UsingNamespace 'Microsoft.Win32.SafeHandles' -ErrorAction SilentlyContinue
+if (-not $NativeMethods) {
+    throw "Fatal error: Unable to compile the necessary C# code."
+}
+
+$ENABLE_VIRTUAL_TERMINAL_PROCESSING = 0x04
+
+function IsVirtualTerminalProcessingEnabled {
+    $ConOut = [Console.NativeMethods]::GetConOut();
+    try {
+        $Mode = 0
+        if ([Console.NativeMethods]::GetConsoleMode($ConOut, [ref] $Mode) -eq 0) {
+            return $false
+        }
+
+        if (-not($Mode -band $ENABLE_VIRTUAL_TERMINAL_PROCESSING)) {
+            return $false
+        }
+
+        return $true
+    } finally {
+        $ConOut.Close()
+    }
+}
+
+function ColorToControlCode{
+    param(
+        [System.Drawing.Color]$Color
+    )
+    if ($IsVirtualTerminalProcessingEnabled) {
+        "`e" + '[38;2;' + ('{0:d3}' -f $Color.R) + ';' + ('{0:d3}' -f $Color.G) + ';' + ('{0:d3}' -f $Color.B) + 'm'
+    }
+}
+
+function ResetForegroundControlCode{
+    if ($IsVirtualTerminalProcessingEnabled) {
+        "`e[39m"
+    }
+}
+
+$global:IsVirtualTerminalProcessingEnabled = IsVirtualTerminalProcessingEnabled
```

Comment:
This compiles the C# when the files in dot-sourced by the module. Which makes the module import time about 0.5 seconds longer... This should be lazier.